src/LocalDatastore.js

```diff
@@ -0,0 +1,210 @@
+    if (!object.objectId) {
+      return;
+    } else {
+      encountered[object.objectId] = object;
+    }
+    for (const key in object) {
+      if (object[key].__type && object[key].__type === 'Object') {
+        this._traverse(object[key], encountered);
+      }
+    }
+  },
+
+  _serializeObjectsFromPinName(name: string) {
+    const localDatastore = this._getLocalDatastore();
+    const allObjects = [];
+    for (const key in localDatastore) {
+      if (key !== DEFAULT_PIN && !key.startsWith(PIN_PREFIX)) {
+        allObjects.push(localDatastore[key]);
+      }
+    }
+    if (!name) {
+      return allObjects;
+    }
+    const pinName = this.getPinName(name);
+    const pinned = this.fromPinWithName(pinName);
+    if (!Array.isArray(pinned)) {
+      return [];
+    }
+    return pinned.map((objectId) => this.fromPinWithName(objectId));
+  },
+
+  _updateObjectIfPinned(object: ParseObject) {
+    if (!this.isEnabled) {
+      return;
+    }
+    const pinned = this.fromPinWithName(object.id);
+    if (pinned) {
+      this.pinWithName(object.id, object._toFullJSON());
+    }
+  },
+
+  _destroyObjectIfPinned(object: ParseObject) {
+    if (!this.isEnabled) {
+      return;
+    }
+    const pin = this.fromPinWithName(object.id);
+    if (!pin) {
+      return;
+    }
+    this.unPinWithName(object.id);
+    const localDatastore = this._getLocalDatastore();
+    for (const key in localDatastore) {
+      if (key === DEFAULT_PIN || key.startsWith(PIN_PREFIX)) {
+        let pinned = this.fromPinWithName(key) || [];
+        if (pinned.includes(object.id)) {
+          pinned = pinned.filter(item => item !== object.id);
+          if (pinned.length == 0) {
+            this.unPinWithName(key);
+          } else {
+            this.pinWithName(key, pinned);
+          }
+        }
+      }
+    }
+  },
+
+  _updateLocalIdForObjectId(localId, objectId) {
+    const unsaved = this.fromPinWithName(localId);
+    if (!unsaved) {
+      return;
+    }
+    this.unPinWithName(localId);
+    this.pinWithName(objectId, unsaved);
+
+    const localDatastore = this._getLocalDatastore();
+
+    for (const key in localDatastore) {
+      if (key === DEFAULT_PIN || key.startsWith(PIN_PREFIX)) {
+        let pinned = this.fromPinWithName(key) || [];
+        if (pinned.includes(localId)) {
+          pinned = pinned.filter(item => item !== localId);
+          pinned.push(objectId);
+          this.pinWithName(key, pinned);
+        }
+      }
+    }
+  },
+
+  getPinName(pinName: ?string) {
+    if (!pinName || pinName === DEFAULT_PIN) {
+      return DEFAULT_PIN;
+    }
+    return PIN_PREFIX + pinName;
+  },
+
+  checkIfEnabled() {
+    if (!this.isEnabled) {
+      console.log('Parse.enableLocalDatastore() must be called first'); // eslint-disable-line no-console
+    }
+    return this.isEnabled;
+  }
+};
+
+function isLocalStorageEnabled() {
+  const item = 'parse_is_localstorage_enabled';
+  try {
+    localStorage.setItem(item, item);
```

Comment:
looks harsh as a test method no? wouldn't testing `typeof localStorage.setItem` and `localStorage.removeItem`
be enough?